File: src/pipeline.js
import { analyze } from "./analysis/index.js"
import { assembly, clambda, cmm, lambda, typedtree } from "./backend/render.js"
import { evaluate, formatValue } from "./runtime/evaluate.js"
import { parse } from "./syntax/parser.js"
import { formatType, infer } from "./type/infer.js"

export const stages = [
  { key: "source", name: "Source", phase: "surface syntax", note: "The parser recognizes functional expressions with literals, collections, constructors and pattern matching." },
  { key: "typedtree", name: "Typedtree", phase: "type checking", note: "Hindley–Milner inference annotates expressions and generalizes let-bound values." },
  { key: "lambda", name: "Lambda", phase: "language lowering", note: "Typed syntax becomes a compact functional intermediate representation." },
  { key: "clambda", name: "Clambda", phase: "closure conversion", note: "Functions expose captured environments and generic closure calls." },
  { key: "cmm", name: "Cmm", phase: "machine lowering", note: "Tagged arithmetic and allocation become explicit machine operations." },
  { key: "assembly", name: "Assembly", phase: "instruction selection", note: "The final sketch shows x86-64 operations over tagged values." }
]

export function compile(source) {
  const ast = infer(parse(source))
  const report = analyze(ast)
  const output = { source: source.trim(), typedtree: typedtree(ast), lambda: lambda(ast), clambda: clambda(ast), cmm: cmm(ast), assembly: assembly(ast) }
  const value = evaluate(ast)
  return { ast, type: formatType(ast.type), nodes: report.metrics.nodes, analysis: report, value, result: formatValue(value), stages: stages.map(stage => ({ ...stage, output: output[stage.key], lines: output[stage.key].split("\n").length })) }
}

export const samples = [
  { name: "tagged arithmetic", source: "let x = 21 in let y = x * 2 in y + 1" },
  { name: "closure capture", source: "let offset = 7 in let add = fun value -> value + offset in add 35" },
  { name: "polymorphic identity", source: "let identity = fun value -> value in let number = identity 42 in identity true" },
  { name: "recursive function", source: "let rec sum n = if n = 0 then 0 else n + sum (n - 1) in sum 6" },
  { name: "list pattern", source: "let head = function | [] -> 0 | value :: rest -> value in head [42; 7]" },
  { name: "option pattern", source: "let unwrap = function | None -> 0 | Some value -> value in unwrap (Some 42)" }
]

File: ocaml/pipeline.ml
type stage = {
  key : string;
  name : string;
  phase : string;
  note : string;
  output : string;
  lines : int;
}

type result = {
  source : string;
  syntax : Syntax.expression;
  inference : Infer.result;
  analysis : Analysis.report;
  value : Value.value;
  stages : stage list;
}

let descriptions =
  [ "source", "Source", "surface syntax",
    "The parser recognizes functional expressions with literals, collections, constructors and pattern matching.";
    "typedtree", "Typedtree", "type checking",
    "Hindley-Milner inference annotates expressions and generalizes let-bound values.";
    "lambda", "Lambda", "language lowering",
    "Typed syntax becomes a compact functional intermediate representation.";
    "clambda", "Clambda", "closure conversion",
    "Functions expose captured environments and generic closure calls.";
    "cmm", "Cmm", "machine lowering",
    "Tagged arithmetic and allocation become explicit machine operations.";
    "assembly", "Assembly", "instruction selection",
    "The final sketch shows x86-64 operations over tagged values." ]

let line_count text =
  if text = "" then 0
  else
    String.fold_left
      (fun count character -> if character = '\n' then count + 1 else count)
      1
      text

let stage outputs (key, name, phase, note) =
  let output = List.assoc key outputs in
  { key; name; phase; note; output; lines = line_count output }

let compile source =
  let syntax = Parser.parse source in
  let inference = Infer.infer syntax in
  let analysis = Analysis.analyze syntax in
  let value = Eval.run syntax in
  let outputs =
    [ "source", String.trim source;
      "typedtree", Lower.typedtree inference syntax;
      "lambda", Lower.lambda syntax;
      "clambda", Lower.clambda syntax;
      "cmm", Lower.cmm syntax;
      "assembly", Lower.assembly syntax ]
  in
  { source;
    syntax;
    inference;
    analysis;
    value;
    stages = List.map (stage outputs) descriptions }

let result_type result =
  Infer.format_result result.inference

let result_value result =
  Value.format result.value

let find_stage result name =
  List.find_opt (fun stage -> stage.key = String.lowercase_ascii name) result.stages

let sample =
  "let rec sum n =\n  if n = 0 then 0\n  else n + sum (n - 1)\nin\nsum 6"
